### PR TITLE
Add custom display names for expressions and constraints

### PR DESCRIPTION
**Pull request by ClayCampaigne**
## Description
Adds set_display_name() and display_name() methods to expressions and constraints, allowing users to assign custom labels for better debugging and model interpretation.

Key features:
- Method chaining: set_display_name() returns self to allow (x >= 0).set_display_name("bounds")
- Hybrid behavior: constraints show display names in str(), expressions don't (prevents component expression display name labels from leaking into compound expression names)
- No breaking changes: all existing code works exactly as before
- Comprehensive testing: 18 tests covering all functionality and edge cases
- Documentation: added to intro tutorial with practical examples

🎉 Generated with [Claude Code](https://claude.ai/code)
Issue link (if applicable): https://github.com/cvxpy/cvxpy/issues/322, https://github.com/cvxpy/cvxpy/issues/2619

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by PTNobel** [comment]
This looks really cool!

**Comment by ClayCampaigne** [comment]
Thanks for the review! Lmk if you want me to pare back some of the new tests. 18 new tests, 333 new lines, seems like overkill for this feature now that I look back at it.

File: cvxpy/constraints/constraint.py [hunk 1]
```diff
@@ -52,14 +52,46 @@ def __init__(self, args, constr_id=None) -> None:
...
+        """
+        self._display_name = str(display_name) if display_name is not None else None
+        return self
+
+    def display_name(self) -> str:
+        """Get the display name of this constraint.
+        
+        Returns custom display name if set, otherwise mathematical form.
+        """
+        if hasattr(self, '_display_name') and self._display_name is not None:
```
Comment (jonathanberthias) [inline cvxpy/constraints/constraint.py:91]: Could we initialize the attribute to `None` in `__init__` to avoid having to check whether it exists every time? Or maybe the default should be an empty string?

File: cvxpy/constraints/constraint.py [hunk 2]
```diff
@@ -52,14 +52,46 @@ def __init__(self, args, constr_id=None) -> None:
...
+        --------
+        >>> x = cp.Variable(3)
+        >>> con = (x >= 0).set_display_name("non_negative")
+        >>> str(con)
+        'non_negative: 0.0 <= x'
+        """
+        self._display_name = str(display_name) if display_name is not None else None
+        return self
+
+    def display_name(self) -> str:
```
Comment (jonathanberthias) [inline cvxpy/constraints/constraint.py:86]: At first, I expected this method to return the set display name, not the name and the stringified constraint. I guess it's due to the infortunate `name` used for the `str` implementation, but maybe we can do better now?
I would suggest having this method return just the display name, whether it has been set or the default, and then put the logic of how to display the constraint inside `__str__` directly. What do you think?

File: cvxpy/constraints/constraint.py [hunk 3]
```diff
@@ -52,14 +52,46 @@ def __init__(self, args, constr_id=None) -> None:
...
-        return self.name()
+        return self.display_name()
 
     def __repr__(self) -> str:
         """Returns a string with information about the constraint.
         """
         return "%s(%s)" % (self.__class__.__name__,
                            repr(self.args[0]))
 
+    def set_display_name(self, display_name: str):
```
Comment (jonathanberthias) [inline cvxpy/constraints/constraint.py:63]: I'm afraid this method's name is too long given how frequently I would use it. For instance, a typical use-case could be:
```py
def build_problem(data):
    weights = cp.Variable(len(data))
    constraints = [
        (weights @ data.values <= data.upper_bound).set_display_name("no over-allocation")
    ]
```
but the constraint line is already too long and `black` or `ruff` formatters would split it which make things very weird:
```py
def build_problem(data):
    weights = cp.Variable(len(data))
    constraints = [
        (weights @ data.values <= data.upper_bound).set_display_name(
            "no over-allocation"
        )
    ]
```

So the simple solution would be to reduce the length of the name. How about `with_name`? Or maybe even just `named`?

File: cvxpy/constraints/zero.py [hunk 4]
```diff
@@ -34,7 +34,7 @@ def __init__(self, expr, constr_id=None) -> None:
     def __str__(self):
         """Returns a string showing the mathematical constraint.
         """
-        return self.name()
+        return self.display_name()
```
Comment (jonathanberthias) [inline cvxpy/constraints/zero.py:37]: I'm pretty sure this override can be removed entirely

File: cvxpy/constraints/zero.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def __init__(self, lhs, rhs, constr_id=None) -> None:
...
```
Comment (jonathanberthias) [inline cvxpy/constraints/zero.py:109]: Same here

File: cvxpy/expressions/expression.py [hunk 6]
```diff
@@ -175,6 +175,38 @@ def name(self) -> str:
...
+        --------
+        >>> x = cp.Variable(3)
+        >>> expr = cp.sum(x).set_display_name("total")
+        >>> expr.display_name()
+        'total: Sum(x, None, False)'
+        """
+        self._display_name = display_name
+        return self
+
+    def display_name(self) -> str:
```
Comment (jonathanberthias) [inline cvxpy/expressions/expression.py:202]: I have the same doubt as for the constraint, even though it's less important to me for expressions since I would have to explicitly call this method anyway.
Still, I wouldn't expect `display_name` to show me more than the name I set

**Review by jonathanberthias — COMMENTED** [review]
Thanks a lot for tackling this! I'm not a maintainer here, but I can definitely see myself using this feature quite a bit, hence my remarks and limitations. It's all up for discussion of course!

File: cvxpy/constraints/constraint.py (hunk 1 above)
Comment (PTNobel) [reply cvxpy/constraints/constraint.py:91]: Agreed that this would be a better approach

File: cvxpy/constraints/constraint.py (hunk 2 above)
Comment (PTNobel) [reply cvxpy/constraints/constraint.py:86]: This makes sense to me

File: cvxpy/constraints/constraint.py (hunk 3 above)
Comment (ClayCampaigne) [reply cvxpy/constraints/constraint.py:63]: another option I considered is `set_label`, which makes clearer that the display name has two parts, the label and the original mathematical formulation of the constraint, which in this design is preserved as a part of the display name.
I'm open to `with_name` or `named` as well. Maybe someone else can weigh in.

File: cvxpy/constraints/constraint.py (hunk 2 above)
Comment (ClayCampaigne) [reply cvxpy/constraints/constraint.py:86]: @jonathanberthias maybe `label` works well as the method name for your suggested approach? Then the idea of my current proposal would be that the constraint displays with either just the original "stringified constraints," and if we have set it, prefixed by the label. 
I kind of like the option to have both a user-set label and the current stringified constraint, but I'm not super strongly attached to it.

File: cvxpy/constraints/zero.py (hunk 4 above)
Comment (ClayCampaigne) [reply cvxpy/constraints/zero.py:37]: Good catch. There is no reason for this child class to implement the same method.

**Comment by ClayCampaigne** [comment]
Thanks for the feedback! I may come back with some design questions on discord before addressing these comments @jonathanberthias @PTNobel

File: cvxpy/constraints/constraint.py (hunk 2 above)
Comment (jonathanberthias) [reply cvxpy/constraints/constraint.py:86]: I really like label 💯 it is much clearer IMO than display name

**Comment by ClayCampaigne** [comment]
I'm going to close this one and replace it with #2935 which is more full-featured. Thanks for the feedback!

File: cvxpy/constraints/constraint.py (hunk 3 above)
Comment (PTNobel) [reply cvxpy/constraints/constraint.py:63]: I like label!
